compiler/src/dotty/tools/dotc/transform/sjs/ExplicitJSClasses.scala

```diff
@@ -12,54 +17,728 @@ import core.Types._
+        wrapWithContextualJSClassValue(sup.symbol.asClass.superClass.typeRef)(tree)
       case _ =>
         tree
     }
   }
+
+  /** Generates the desugared version of `js.constructorOf[tpe]`.
+   *
+   *  This is the meat of step (F).
+   */
+  private def genJSConstructorOf(tree: Tree, tpe0: Type)(using Context): Tree = {
+    val tpe = tpe0.underlyingClassRef(refinementOK = false) match {
+      case typeRef: TypeRef => typeRef
+      case _ =>
+        // This should not have passed the checks in PrepJSInterop
+        report.error(i"class type required but found $tpe0", tree)
+        jsdefn.JSObjectType
+    }
+    val cls = tpe.typeSymbol
+
+    // This should not have passed the checks in PrepJSInterop
+    assert(!cls.isOneOf(Trait | ModuleClass),
+        i"non-trait class type required but $tpe found for genJSConstructorOf at ${tree.sourcePos}")
+
+    if (isInnerJSClass(cls)) {
+      // Use the $jsclass field in the outer instance
+      val prefix: Type = tpe.prefix
+      if (prefix.isStable) {
+        val jsclassAccessor = jsclassAccessorFor(cls)
+        ref(NamedType(prefix, jsclassAccessor.name, jsclassAccessor.denot))
+      } else {
+        report.error(i"stable reference to a JS class required but $tpe found", tree)
+        ref(defn.Predef_undefined)
+      }
+    } else if (isLocalJSClass(cls)) {
+      // Use the local `val` that stores the JS class value
+      val state = myState
+      val jsclassVal = state.localClass2jsclassVal(cls)
+      state.notYetSelfReferencingLocalClasses -= cls
+      ref(jsclassVal)
+    } else {
+      // Defer translation to `LoadJSConstructor` to the back-end
+      ref(jsdefn.Runtime_constructorOf).appliedTo(clsOf(tpe))
+    }
+  }
+
+  private def wrapWithContextualJSClassValue(jsClassType: Type)(tree: Tree)(using Context): Tree =
+    wrapWithContextualJSClassValue(genJSConstructorOf(tree, jsClassType))(tree)
+
+  private def wrapWithContextualJSClassValue(jsClassValue: Tree)(tree: Tree)(using Context): Tree =
+    ref(jsdefn.Runtime_withContextualJSClassValue).appliedToType(tree.tpe).appliedTo(jsClassValue, tree)
+
+  private def unwrapWithContextualJSClassValue(tree: Tree)(using Context): Tree = tree match {
+    case Apply(fun, jsClassValue :: actualTree :: Nil)
+        if fun.symbol == jsdefn.Runtime_withContextualJSClassValue =>
+      actualTree
+    case _ =>
+      tree
+  }
+
+  /** Extracts the super type constructor of a `Template`, without type
+   *  parameters, so that the type is well-formed outside of the `Template`,
+   *  i.e., at the same level where the corresponding `TypeDef` is defined.
+   *  It is not necessarily *-kinded, though, which limits its applicability.
+   */
+  private def extractSuperTpeFromImpl(impl: Template)(using Context): Type = {
+    // TODO Check whether stripPoly is the right thing. Do we need a sort of rawTypeRef?
+    impl.parents.head.tpe.stripPoly
```

Comment:
Is this correct?

Reply:
Basically what I'm trying to do here is, given the following code:
```scala
{ // some block/template/package
  class Foo[...Ts] extends pre.Parent[...Us](...args)
  <rest>
}
```
I want to get something that I can put in a `classOf` anywhere in the `<rest>` (so in siblings and nephews of the `TypeDef` node), for example:
```scala
{ // some block/template/package
  class Foo[...Ts] extends pre.Parent[...Us](...args)
  val c = classOf[pre.Parent]
}
```
So here I'm trying to remove the type parameters `[...Us]` from `pre.Parent[...Us]` so that I'm left with `pre.Parent`.

(In scalac I had to actually come up with new type parameters, which involved doing existential abstraction (https://github.com/scala-js/scala-js/blob/ea6a89dabcfa140e7df2441eee69ca34a2dda0f5/compiler/src/main/scala/org/scalajs/nscplugin/JSGlobalAddons.scala#L40-L82), but apparently in dotc I can get away with a bare `TypeRef`.)

Reply:
> So here I'm trying to remove the type parameters [...Us] from pre.Parent[...Us] so that I'm left with pre.Parent.

That should be an `AppliedType(pre.Parent,List(U))`, there's `typeConstructor` to strip that (might want to do a dealias first)